src/scripts/dev_tools/gen_ec_groups.py

```diff
@@ -115,32 +107,100 @@ def format_names(names):
-        order = format_int(order)
-        if order in orders_seen:
-            raise Exception("Duplicate EC group order %s" % (order))
-        orders_seen.add(order)
-        oid = oid[0].replace('.', ', ')
-        yield template_str % (low_bits, order, oid)
+    def __call__(self, l):
+        r = not self.first_line
+        self.first_line = False
+        return r
 
 def main():
     curves = [c for c in curve_info(open('./src/build-data/ec_groups.txt'))]
 
-    template_str = open('./src/build-data/ec_named.cpp.in').read()
+    pcurves = []
+    for c in curves:
+        if 'pcurve' in c['Impl']:
+            pcurves.append(c)
 
-    names = "\n".join(format_names([(c['Name'], c['Impl']) for c in curves]))
-    orders = "\n".join(format_orders([(c['N'], c['OID']) for c in curves]))
-    curves = '\n'.join([print_curve(curve) for curve in curves])
     this_script = sys.argv[0]
-    today = datetime.date.today().strftime("%Y-%m-%d")
+    date = datetime.date.today().strftime("%Y-%m-%d")
+
+    env = Environment(loader=FileSystemLoader("src/build-data/templates"))
+
+    # write ec_named.cpp
+    with open('./src/lib/pubkey/ec_group/ec_named.cpp', encoding='utf8', mode='w') as ec_named:
+        template = env.get_template("ec_named.cpp.in")
+
+        named_groups = "\n".join(format_names([(c['Name'], c['Impl']) for c in curves]))
+
+        ec_named.write(template.render(script=this_script, date=date, curves=curves, named_groups=named_groups.strip()))
+        ec_named.write("\n")
+
+    # write pcurves_instance.h
+    with open('./src/lib/math/pcurves/pcurves_instance.h', encoding='utf8', mode='w') as pcurves_h:
+        template = env.get_template("pcurves_instance.h.in")
+        pcurves_h.write(template.render(script=this_script, date=date, pcurves=pcurves))
+        pcurves_h.write("\n")
+
+    # write pcurves.cpp
+    with open('./src/lib/math/pcurves/pcurves.cpp', encoding='utf8', mode='w') as pcurves_cpp:
+        template = env.get_template("pcurves.cpp.in")
+        pcurves_cpp.write(template.render(script=this_script, date=date, pcurves=pcurves))
+        pcurves_cpp.write("\n")
+
+    # Check if any pcurves modules need a new stub impl
+    for pcurve in pcurves:
+        curve = pcurve["Name"]
+        mod_dir = './src/lib/math/pcurves/pcurves_%s' % (curve)
+        info_path = os.path.join(mod_dir, 'info.txt')
+        impl_path = os.path.join(mod_dir, f'pcurves_{curve}.cpp')
+
+        if os.access(impl_path, os.R_OK):
+            continue
+
+        addchain_fe2 = addchain_code(pcurve['P'] - 3, 0)
+        addchain_scalar = addchain_code(pcurve['N'] - 2, 0)
+
+        try:
+            os.makedirs(mod_dir)
+        except OSError as ex:
+            if ex.errno != errno.EEXIST:
+                raise
+
+        with open(info_path, 'w', encoding='utf8') as info_file:
+            info_file.write(dedent(f"""\
+                <defines>
+                PCURVES_{curve.upper()} -> {datestamp()}
```

Comment:
If the provided EC curve name cannot be transformed into [a valid module preprocessor definition](https://github.com/randombit/botan/blob/b3c5e866bdbd428c239c631e7988a02f0bd61225/configure.py#L965), we should error out. Otherwise `./configure.py` will fail ungracefully. (I ran into this by chance while manually trying the documented flow.)

```suggestion
        module_define = f"PCURVES_{curve.upper()}"
        if not re.match('^[0-9A-Za-z_]{3,30}$', module_define):
            raise ValueError(f"Invalid preprocessor define name ({module_define}) for new pcurve module")

        with open(info_path, 'w', encoding='utf8') as info_file:
            info_file.write(dedent(f"""\
                <defines>
                {module_define} -> {datestamp()}
```